File: apps/api/src/main/java/az/ideanest/shared/observability/Correlation.java
package az.ideanest.shared.observability;

import az.ideanest.shared.Identifiers;
import java.util.Locale;
import java.util.UUID;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.slf4j.MDC;

/**
 * The identifiers §18.1 puts on every line, and the rules for accepting one from
 * outside.
 *
 * <p>The MDC key names are the ones the specification names, so that a query
 * written against the documentation works against the logs.
 *
 * <p><strong>An inbound identifier is input, not information.</strong> A client
 * sends {@code X-Request-Id} so that its own logs and ours can be joined, which
 * is worth honouring — but the value ends up in every line of the request, so
 * echoing it unchecked would let anybody with an HTTP client write whatever they
 * liked into the log stream. A newline forges a log entry; a megabyte forges an
 * outage. Anything outside {@link #ACCEPTABLE_IDENTIFIER} is therefore dropped
 * and replaced by one of ours rather than sanitised into something that no longer
 * joins to anything.
 */
public final class Correlation {

    /** §18.1's five fields. */
    public static final String REQUEST_ID = "requestId";

    public static final String TRACE_ID = "traceId";
    public static final String SPAN_ID = "spanId";
    public static final String USER_ID = "userId";
    public static final String PROJECT_ID = "projectId";

    /** What a client sends, and what comes back on the response. */
    public static final String REQUEST_ID_HEADER = "X-Request-Id";

    public static final String TRACE_ID_HEADER = "X-Trace-Id";

    /** W3C Trace Context. Lower case, per the specification. */
    public static final String TRACEPARENT_HEADER = "traceparent";

    /**
     * Long enough for a UUID with its hyphens, and for the identifiers other
     * platforms mint. Short enough that the header cannot become the payload.
     */
    private static final Pattern ACCEPTABLE_IDENTIFIER = Pattern.compile("[A-Za-z0-9_-]{8,64}");

    /**
     * {@code version-traceid-spanid-flags}. Version {@code ff} is invalid by the
     * specification, an all-zero trace or span identifier likewise. Trailing
     * fields from a future version are tolerated: this reads the two values it
     * needs and ignores the rest, which is what the specification asks of a
     * participant that does not understand a newer version.
     */
    private static final Pattern TRACEPARENT = Pattern.compile(
            "^(?!ff)[0-9a-f]{2}-(?!0{32})([0-9a-f]{32})-(?!0{16})([0-9a-f]{16})-[0-9a-f]{2}(?:-.*)?$");

    /** A header longer than this is not parsed at all. */
    private static final int LONGEST_ACCEPTED_HEADER = 256;

    private Correlation() {
    }

    /** The candidate if it is one this service is willing to write to a log, otherwise null. */
    public static String acceptableIdentifier(String candidate) {
        if (candidate == null || candidate.length() > LONGEST_ACCEPTED_HEADER) {
            return null;
        }
        return ACCEPTABLE_IDENTIFIER.matcher(candidate).matches() ? candidate : null;
    }

    /** The caller's trace, if the header really is one, otherwise null. */
    public static String traceIdFrom(String traceparent) {
        if (traceparent == null || traceparent.length() > LONGEST_ACCEPTED_HEADER) {
            return null;
        }
        Matcher matcher = TRACEPARENT.matcher(traceparent.trim().toLowerCase(Locale.ROOT));
        return matcher.matches() ? matcher.group(1) : null;
    }

    /**
     * A new request identifier: UUID v7, like every other identifier in the
     * service, so that a log line sorts the way the rows it describes do.
     */
    public static String newRequestId() {
        return Identifiers.newIdentifier().toString();
    }

    /** Thirty-two lower-case hex characters, as W3C Trace Context requires. */
    public static String newTraceId() {
        return Identifiers.newIdentifier().toString().replace("-", "");
    }

    /** Sixteen lower-case hex characters. Never the all-zero span. */
    public static String newSpanId() {
        UUID source = Identifiers.newIdentifier();
        long bits = source.getLeastSignificantBits();
        return String.format("%016x", bits == 0 ? 1 : bits);
    }

    /**
     * Puts a fresh set into the MDC, overwriting whatever was there.
     *
     * <p>For work that did not arrive in a request — a scheduled sweep, say.
     * There is no caller to inherit from, and §18.1 wants the line correlated
     * anyway: without this, the one place a job's lines could be gathered from is
     * the timestamp.
     */
    public static void mintIntoMdc() {
        MDC.put(REQUEST_ID, newRequestId());
        MDC.put(TRACE_ID, newTraceId());
        MDC.put(SPAN_ID, newSpanId());
    }
}

File: apps/web/src/lib/rum/correlation.ts
/**
 * The browser half of `az.ideanest.shared.observability.Correlation`.
 *
 * §18.1 puts `requestId`, `traceId` and `spanId` on every line the service logs,
 * and accepts the first from `X-Request-Id` and the second from a W3C
 * `traceparent` when they are shaped acceptably. A field measurement that used a
 * vocabulary of its own would be a second set of identifiers that join to
 * nothing: the whole value of "this session's LCP was nine seconds" is being able
 * to open the trace of the requests that session made.
 *
 * **The shapes below are copied from the Java, deliberately and exactly.** An
 * identifier this file mints and that file rejects is worse than no identifier,
 * because the beacon still carries one and the log still prints it, and nobody
 * discovers it joins to nothing until they try. `correlation.test.ts` pins each
 * pattern against the Java source's own literal.
 *
 * <h2>Why the identifiers travel in the body and not in headers</h2>
 *
 * `navigator.sendBeacon` cannot set a request header. It takes a URL and a body,
 * and the only header it will emit is the `Content-Type` of the `Blob` it is
 * given. So the correlation identifiers are fields of the payload, and the
 * endpoint reads them from there. When the `fetch` fallback runs it sets
 * `X-Request-Id` and `traceparent` as well — same values, so a proxy log and the
 * application log agree — and the endpoint prefers the headers when it has them,
 * exactly as `CorrelationFilter` does.
 */

/** What the service reads a client's request identifier from. */
export const REQUEST_ID_HEADER = 'X-Request-Id';

/** What it returns its own on. */
export const TRACE_ID_HEADER = 'X-Trace-Id';

/** W3C Trace Context. Lower case, per the specification. */
export const TRACEPARENT_HEADER = 'traceparent';

/**
 * `Correlation.ACCEPTABLE_IDENTIFIER`. Long enough for a UUID with its hyphens,
 * short enough that the header cannot become the payload.
 */
const ACCEPTABLE_IDENTIFIER = /^[A-Za-z0-9_-]{8,64}$/;

/** `Correlation.TRACEPARENT`: `version-traceid-spanid-flags`. */
const TRACEPARENT =
  /^(?!ff)[0-9a-f]{2}-(?!0{32})([0-9a-f]{32})-(?!0{16})([0-9a-f]{16})-[0-9a-f]{2}(?:-.*)?$/;

/** `Correlation.LONGEST_ACCEPTED_HEADER`. A longer header is not parsed at all. */
export const LONGEST_ACCEPTED_HEADER = 256;

/** Thirty-two lower-case hex characters, and never the all-zero trace. */
const TRACE_ID = /^(?!0{32})[0-9a-f]{32}$/;

/** Sixteen, and never the all-zero span. */
const SPAN_ID = /^(?!0{16})[0-9a-f]{16}$/;

/**
 * `crypto.randomUUID()`'s output, and nothing looser.
 *
 * A session identifier is the sampling key and the join key, so it has to be
 * *some* stable string; making it a v4 UUID rather than "any acceptable
 * identifier" means it carries 122 bits of `crypto`-grade randomness and cannot
 * be anything a caller chose to put there — an attacker who could pick session
 * identifiers could pick ones that sample in and drown a route's percentile.
 */
const SESSION_ID = /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/;

/** The candidate if the service would accept it as a request identifier, else null. */
export function acceptableIdentifier(candidate: string | null | undefined): string | null {
  if (typeof candidate !== 'string' || candidate.length > LONGEST_ACCEPTED_HEADER) return null;
  return ACCEPTABLE_IDENTIFIER.test(candidate) ? candidate : null;
}

/** The caller's trace, if the header really is one, otherwise null. */
export function traceIdFrom(traceparent: string | null | undefined): string | null {
  if (typeof traceparent !== 'string' || traceparent.length > LONGEST_ACCEPTED_HEADER) return null;
  return TRACEPARENT.exec(traceparent.trim().toLowerCase())?.[1] ?? null;
}

export function isTraceId(candidate: string): boolean {
  return TRACE_ID.test(candidate);
}

export function isSpanId(candidate: string): boolean {
  return SPAN_ID.test(candidate);
}

export function isSessionId(candidate: string): boolean {
  return SESSION_ID.test(candidate);
}

/**
 * The `traceparent` a beacon or an API call announces itself with.
 *
 * Version `00`, and the sampled flag set — a trace we chose to report is one we
 * are asking the service to record spans for. The service continues the trace
 * and mints its own span, so the span here is the parent of whatever it does.
 */
export function traceparentOf(traceId: string, spanId: string): string {
  return `00-${traceId}-${spanId}-01`;
}

type RandomSource = Pick<Crypto, 'getRandomValues'>;

function randomHex(bytes: number, source: RandomSource): string {
  const buffer = new Uint8Array(bytes);
  source.getRandomValues(buffer);
  let hex = '';
  for (const byte of buffer) hex += byte.toString(16).padStart(2, '0');
  return hex;
}

/** Thirty-two lower-case hex characters. Never the all-zero trace. */
export function newTraceId(source: RandomSource = crypto): string {
  const candidate = randomHex(16, source);
  return TRACE_ID.test(candidate) ? candidate : `1${candidate.slice(1)}`;
}

/** Sixteen lower-case hex characters. Never the all-zero span. */
export function newSpanId(source: RandomSource = crypto): string {
  const candidate = randomHex(8, source);
  return SPAN_ID.test(candidate) ? candidate : `1${candidate.slice(1)}`;
}

/**
 * A request identifier for one beacon.
 *
 * A UUID, which `ACCEPTABLE_IDENTIFIER` admits — the service mints UUID v7 for
 * the same field and a browser has no v7 generator, but the shape the service
 * validates is the same and the sort order of a client-minted identifier buys
 * nothing on the server side anyway.
 */
export function newRequestId(random: () => string = () => crypto.randomUUID()): string {
  return random();
}
